File: C/3/2.c
#include <stdio.h>

/* Defining a maximum size to strings */
#define SIZE 999

/**
 * A function that converts characters like newline and tab into visible escape
 * sequences like \n and \t as it copies the string t to s.
 *
 * @param char[] s String to be copied to
 * @param char[] t Source string
 *
 * @return void
 */
void escape(char s[], char t[])
{
    int i = 0;
    int j = 0;
    char c;

    while ((c = t[i++]) != '\0') {
        switch (c) {
            case '\n':
                s[j++] = '\\';
                s[j++] = '\\';
                s[j++] = 'n';
                break;
            case '\t':
                s[j++] = '\\';
                s[j++] = '\\';
                s[j++] = 't';
                break;
            default:
                s[j++] = c;
                break;
        }
    }

    /* Adding null character */
    s[j] = '\0';

    return;
}

/**
 * A function that converts visible escape sequences like \n and \t into
 * characters like newline and tab as it copies the string t to s.
 *
 * @param char[] s String to be copied to
 * @param char[] t Source string
 *
 * @return void
 */
void escape_back(char s[], char t[])
{
    int i = 0;
    int j = 0;
    char c;

    while ((c = t[i++]) != '\0') {
        switch (c) {
            case '\\':
                if (t[i] == '\\' && t[i+1] == 'n') {
                    s[j++] = '\n';
                } else if (t[i] == '\\' && t[i+1] == 't') {
                    s[j++] = '\t';
                }
                i += 2;
                break;
            default:
                s[j++] = c;
                break;
        }
    }

    /* Adding null character */
    s[j] = '\0';

    return;
}

/**
 * A program to test string copying.
 * For the sake of simplicity, we assume only '\t' and '\n' can be scaped.
 */
main()
{
    char input1[SIZE] = "A \n test \t whatever.";
    char output1[SIZE];
    char output2[SIZE];

    escape(output1, input1);

    printf("escape('%s') => '%s'\n", input1, output1);

    escape_back(output2, output1);

    printf("escape_back('%s') => '%s'\n", output1, output2);
}

File: rustlang/3/2.rs
/// A function that converts characters like newline and tab into visible escape
/// sequences like \n and \t as it copies the string t to s.
///
/// @param  &str s Source string
///
/// @return String Scaped string
fn escape(s : &str) -> String
{
    let mut result = String::with_capacity(s.len());

    for c in s.chars() {
        match c {
            '\n' => {
                result.push('\\');
                result.push('\\');
                result.push('n');
            },
            '\t' => {
                result.push('\\');
                result.push('\\');
                result.push('t');
            },
            _ => {
                result.push(c);
            }
        }
    }

    result
}

/// A function that converts visible escape sequences like \n and \t into
/// characters like newline and tab as it copies the string t to s.
///
/// @param  &str s Source string
///
/// @return String (UN)escaped string
fn escape_back(s : &str) -> String
{
    let mut result = String::with_capacity(s.len());

    // We use a peekable iterator, so we can check characters ahead
    let mut iter = s.chars().peekable();

    loop {
        let mut _c = iter.next();

        match _c {
            None => {
                // There are no more bytes to be read
                break;
            },
            Some('\\') => {
                if iter.peek() == Some(&&'\\'){
                    _c = iter.next();
                }

                if iter.peek() == Some(&&'n'){
                    result.push('\n');
                } else if iter.peek() == Some(&&'t'){
                    result.push('\t');
                }
                iter.next();
            }
            _ => {
                result.push(_c.unwrap());
            }
        }
    }

    result
}

/// A program to test string copying.
/// For the sake of simplicity, we assume only '\t' and '\n' can be scaped.
fn main()
{
    let input1 = String::from("A \n test \t whatever.");

    let output1 = escape(&input1);

    print!("escape('{}') => '{}'\n", input1, output1);

    let output2 = escape_back(&output1);

    print!("escape_back('{}') => '{}'\n", output1, output2);
}
